{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run AttnCNP Large\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/private/home/yannd/projects/Neural-Process-Family\n"
     ]
    }
   ],
   "source": [
    "cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.environ['CUDA_VISIBLE_DEVICES'] = \"1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import neuralproc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import submitit\n",
    "from train_imgs import main, parse_arguments\n",
    "\n",
    "log_folder=\"logs/%j\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def repeat_list(l, n):\n",
    "    return [i for i in l for _ in range(n)]\n",
    "\n",
    "class Run:\n",
    "    def checkpoint(self, args):\n",
    "        \"\"\"Resubmits the same callable with the same arguments but makes sure continnue from last chckpnt.\"\"\"\n",
    "        args.is_continue_train=True \n",
    "        return submitit.utils.DelayedSubmission(self, args)\n",
    "    \n",
    "    def __call__(self, args):\n",
    "        job_env = submitit.utils.JobEnvironment()\n",
    "        args.starting_run = args.starting_run * job_env.num_tasks + job_env.local_rank\n",
    "        return main(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = [\"celeba32\", \"celeba64\", \"svhn\", \"mnist\", \"zs-multi-mnist\"]\n",
    "\n",
    "args = [parse_arguments(\"{} {} --starting-run {} --batch-size 8 --patience 5 --is-progressbar\".format(m, d, s).split())\n",
    "        for s in range(6)\n",
    "        for m in [\"SelfAttnCNP\"] \n",
    "        for d in [\"zs-multi-mnist\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "executor=submitit.SlurmExecutor(folder=log_folder, max_num_timeout=3)\n",
    "executor.update_parameters(num_gpus=1, \n",
    "                           time=6,#0*24*3, \n",
    "                           partition=\"dev\", \n",
    "                           cpus_per_task=10, \n",
    "                           mem='32GB',\n",
    "                           constraint=\"volta32gb\"\n",
    "                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "job = executor.map_array(Run(), args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SlurmJob<job_id=17816174_0, task_id=0, state=\"CANCELLED+\">,\n",
       " SlurmJob<job_id=17816174_1, task_id=0, state=\"CANCELLED+\">,\n",
       " SlurmJob<job_id=17816174_2, task_id=0, state=\"CANCELLED+\">,\n",
       " SlurmJob<job_id=17816174_3, task_id=0, state=\"CANCELLED+\">,\n",
       " SlurmJob<job_id=17816174_4, task_id=0, state=\"CANCELLED+\">,\n",
       " SlurmJob<job_id=17816174_5, task_id=0, state=\"CANCELLED+\">]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "j = job[0]\n",
    "job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "for j in job:\n",
    "    j.cancel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:29,634) - Starting with JobEnvironment(job_id=17816135_0, hostname=learnfair0339, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:29,634) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_0/17816135_0_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_0 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:54:47,139) - Submitted job triggered an exception\n",
      "\n",
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:29,422) - Starting with JobEnvironment(job_id=17816135_1, hostname=learnfair0341, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:29,423) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_1/17816135_1_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_1 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:54:46,258) - Submitted job triggered an exception\n",
      "\n",
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:29,633) - Starting with JobEnvironment(job_id=17816135_2, hostname=learnfair0363, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:29,633) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_2/17816135_2_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_2 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:54:45,848) - Submitted job triggered an exception\n",
      "\n",
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:29,633) - Starting with JobEnvironment(job_id=17816135_3, hostname=learnfair0363, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:29,633) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_3/17816135_3_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_3 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:54:45,605) - Submitted job triggered an exception\n",
      "\n",
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:30,847) - Starting with JobEnvironment(job_id=17816135_4, hostname=learnfair0368, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:30,847) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_4/17816135_4_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_4 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:55:02,131) - Submitted job triggered an exception\n",
      "\n",
      "--------------------------------\n",
      "submitit INFO (2019-09-18 09:54:30,847) - Starting with JobEnvironment(job_id=17816135_5, hostname=learnfair0368, local_rank=0(1), node=0(1), global_rank=0(1))\n",
      "submitit INFO (2019-09-18 09:54:30,847) - Loading pickle: /private/home/yannd/projects/Neural-Process-Family/logs/17816135_5/17816135_5_submitted.pkl\n",
      "\n",
      "--- Training zs-multi-mnist/SelfAttnCNP/run_5 ---\n",
      "\n",
      "submitit ERROR (2019-09-18 09:55:02,132) - Submitted job triggered an exception\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for j in job:\n",
    "    print(\"--------------------------------\")\n",
    "    print(j.stdout())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:54:47,139) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 611, in train_step_single\n",
      "    loss.backward()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/tensor.py\", line 118, in backward\n",
      "    torch.autograd.backward(self, gradient, retain_graph, create_graph)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/autograd/__init__.py\", line 93, in backward\n",
      "    allow_unreachable=True)  # allow_unreachable flag\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 4.69 GiB (GPU 0; 31.72 GiB total capacity; 24.00 GiB already allocated; 2.22 GiB free; 4.48 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0339: task 0: Exited with exit code 1\n",
      "\n",
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:54:46,258) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 611, in train_step_single\n",
      "    loss.backward()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/tensor.py\", line 118, in backward\n",
      "    torch.autograd.backward(self, gradient, retain_graph, create_graph)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/autograd/__init__.py\", line 93, in backward\n",
      "    allow_unreachable=True)  # allow_unreachable flag\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 4.69 GiB (GPU 0; 31.72 GiB total capacity; 24.74 GiB already allocated; 1.46 GiB free; 4.49 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0341: task 0: Exited with exit code 1\n",
      "\n",
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:54:45,848) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 611, in train_step_single\n",
      "    loss.backward()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/tensor.py\", line 118, in backward\n",
      "    torch.autograd.backward(self, gradient, retain_graph, create_graph)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/autograd/__init__.py\", line 93, in backward\n",
      "    allow_unreachable=True)  # allow_unreachable flag\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 4.69 GiB (GPU 0; 31.72 GiB total capacity; 24.57 GiB already allocated; 1.60 GiB free; 4.52 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0363: task 0: Exited with exit code 1\n",
      "\n",
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:54:45,605) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 609, in train_step_single\n",
      "    y_pred = self.infer(Xi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 978, in infer\n",
      "    return self.module_(**x_dict)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/nn/modules/module.py\", line 547, in __call__\n",
      "    result = self.forward(*input, **kwargs)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/neuralproc.py\", line 278, in forward\n",
      "    R_det = self.deterministic_path(X_cntxt, Y_cntxt, X_trgt)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/neuralproc.py\", line 454, in deterministic_path\n",
      "    R_attn = self.attender(keys, queries, values, **attender_kwargs)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/nn/modules/module.py\", line 547, in __call__\n",
      "    result = self.forward(*input, **kwargs)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/utils/attention.py\", line 555, in forward\n",
      "    context = super().forward(keys, queries, values, **kwargs)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/utils/attention.py\", line 477, in forward\n",
      "    context = self.dot(keys, queries, values)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/nn/modules/module.py\", line 547, in __call__\n",
      "    result = self.forward(*input, **kwargs)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/utils/attention.py\", line 145, in forward\n",
      "    logits = self.score(keys, queries, **kwargs)\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/neuralproc/utils/attention.py\", line 219, in score\n",
      "    logits = logits / math.sqrt(kq_size)\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 1.80 GiB (GPU 0; 31.72 GiB total capacity; 27.32 GiB already allocated; 1.19 GiB free; 2.18 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0363: task 0: Exited with exit code 1\n",
      "\n",
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:55:02,131) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 611, in train_step_single\n",
      "    loss.backward()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/tensor.py\", line 118, in backward\n",
      "    torch.autograd.backward(self, gradient, retain_graph, create_graph)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/autograd/__init__.py\", line 93, in backward\n",
      "    allow_unreachable=True)  # allow_unreachable flag\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 4.69 GiB (GPU 0; 31.72 GiB total capacity; 23.98 GiB already allocated; 2.20 GiB free; 4.51 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0368: task 0: Exited with exit code 1\n",
      "\n",
      "--------------------------------\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "INFO:utils.data.imgs:Resizing ZeroShotMultiMNIST ...\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=128 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "/private/home/yannd/projects/Neural-Process-Family/neuralproc/predefined/mlp.py:67: UserWarning: hidden_size=32 smaller than output=256 and input=128. Setting it to 128.\n",
      "  warnings.warn(txt.format(hidden_size, output_size, input_size, self.hidden_size))\n",
      "\n",
      "  0%|          | 0/3563 [00:00<?, ?it/s]submitit ERROR (2019-09-18 09:55:02,132) - Submitted job triggered an exception\n",
      "ERROR:submitit:Submitted job triggered an exception\n",
      "Traceback (most recent call last):\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/_submit.py\", line 6, in <module>\n",
      "    submitit_main()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 186, in submitit_main\n",
      "    process_job(args.folder)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 179, in process_job\n",
      "    raise error\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/submission.py\", line 168, in process_job\n",
      "    result = delayed.result()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/submitit/utils.py\", line 122, in result\n",
      "    self._result = self.function(*self.args, **self.kwargs)\n",
      "  File \"<ipython-input-5-21e536907d9e>\", line 13, in __call__\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 216, in main\n",
      "    chckpnt_dirname=args.chckpnt_dirname,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/train_imgs.py\", line 181, in train\n",
      "    **kwargs,\n",
      "  File \"/private/home/yannd/projects/Neural-Process-Family/utils/train.py\", line 216, in train_models\n",
      "    _ = trainer.fit(data_train)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 852, in fit\n",
      "    self.partial_fit(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 811, in partial_fit\n",
      "    self.fit_loop(X, y, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 743, in fit_loop\n",
      "    step = self.train_step(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 670, in train_step\n",
      "    self.optimizer_.step(step_fn)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/optim/adam.py\", line 58, in step\n",
      "    loss = closure()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 667, in step_fn\n",
      "    step = self.train_step_single(Xi, yi, **fit_params)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/skorch/net.py\", line 611, in train_step_single\n",
      "    loss.backward()\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/tensor.py\", line 118, in backward\n",
      "    torch.autograd.backward(self, gradient, retain_graph, create_graph)\n",
      "  File \"/private/home/yannd/.conda/envs/neuralproc/lib/python3.7/site-packages/torch/autograd/__init__.py\", line 93, in backward\n",
      "    allow_unreachable=True)  # allow_unreachable flag\n",
      "RuntimeError: CUDA out of memory. Tried to allocate 4.69 GiB (GPU 0; 31.72 GiB total capacity; 24.54 GiB already allocated; 1.65 GiB free; 4.50 GiB cached)\n",
      "\n",
      "                                        \n",
      "srun: error: learnfair0368: task 0: Exited with exit code 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for j in job:\n",
    "    print(\"--------------------------------\")\n",
    "    print(j.stderr())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
